REMEMBER: FIRST CREATE A COPY OF THIS FILE WITH A UNIQUE NAME AND DO YOUR WORK THERE. AND MAKE SURE YOU COMMIT YOUR CHANGES TO THE `hw3_submissions` BRANCH.

# Assignment 3 | Cleaning and Exploring Data with Pandas



<img src="data/scoreCard.jpg" width=250>

In this assignment, you will investigate restaurant food safety scores for restaurants in San Francisco. Above is a sample score card for a restaurant. The scores and violation information have been made available by the San Francisco Department of Public Health. 

## Loading Food Safety Data


There are 2 files in the data directory:
1. business.csv containing food establishments in San Francisco
1. inspections.csv containing retaurant inspections records

Let's start by loading them into Pandas dataframes.  One of the files, business.csv, has encoding (ISO-8859-1), so you will need to account for that when reading it.

### Question 1

#### Question 1a
Read the two files noted above into two pandas dataframes named `bus` and `ins`, respectively. Print the first 5 rows of each to inspect them.


In [1]:
import pandas as pd

In [2]:
#Do do this, I had to first go into the terminal and do the following code once in the a3 folder: unzip 'data.zip'
bus = pd.read_csv('data/businesses.csv', encoding='ISO-8859-1')
ins = pd.read_csv('data/inspections.csv')

In [3]:
bus.head()

business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           31      NORMAN'S ICE CREAM AND FREEZES   
4           45                 CHARLIE'S DELI CAFE   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   
4                3202 FOLSOM ST   San Francisco    CA       94110  37.747114   

    longitude  phone_number  
0 -122.403816  +14154217044  
1 -122.421547  +14157763262  
2 -122.403135  +14156779494  
3 -122.419004           NaN  
4 -122.413641  +14156415051

In [4]:
ins.head()

business_id  score      date     type
0           10     82  20160503  routine
1           10     94  20140729  routine
2           10     92  20140114  routine
3           19     94  20160513  routine
4           19     94  20141110  routine

## Examining the Business data

From its name alone, we expect the `businesses.csv` file to contain information about the restaurants. Let's investigate this dataset.

### Question 2

#### Question 2a: How many records are there?

In [5]:
bus.describe() #prints a summary of the data, but does not necessarily show how many rows/records b/c there could be 
                #duplicates of IDs

business_id     latitude    longitude
count   6315.000000  3599.000000  3599.000000
mean   48562.416627    37.773807  -122.425298
std    33825.711496     0.022876     0.027568
min       10.000000    37.668824  -122.510896
25%     5891.500000    37.760571  -122.436385
50%    65012.000000    37.780739  -122.418626
75%    79197.000000    37.790038  -122.406149
max    90239.000000    37.834628  -122.264171

In [6]:
#prints length of the csv

print("number of lines in businesses csv: ",len(bus))

number of lines in businesses csv:  6315


#### Question 2b: How many unique business IDs are there?  

In [7]:
#print unique for ONLY business ID column by creating new variable
unique = len(pd.unique(bus['business_id']))

print("The number of unique business ids: ", unique)

The number of unique business ids:  6315


In [8]:
unique2 = bus.nunique(axis = 0) #prints number of unique for each column too
print(unique2)

business_id     6315
name            5692
address         5534
city               1
state              1
postal_code       46
latitude        2865
longitude       2914
phone_number    4919
dtype: int64


#### Question 2c: What are the 5 most common businesses by name, and how many are there in San Francisco?

In [9]:
#gets most common 

bus['name'].mode()  

0    STARBUCKS COFFEE
dtype: object

In [10]:
#another way to return top 5 in a list

top5 = 5
bus['name'].value_counts()[:top5].index.tolist()

['STARBUCKS COFFEE',
 "PEET'S COFFEE & TEA",
 'MCDONALDS',
 'SAN FRANCISCO SOUP COMPANY',
 'WALGREENS']

In [11]:
#a third way to get top 5 and the number of times they occur. 

bus['name'].value_counts()[0:5]


STARBUCKS COFFEE              72
PEET'S COFFEE & TEA           24
MCDONALDS                     12
SAN FRANCISCO SOUP COMPANY    11
WALGREENS                     11
Name: name, dtype: int64

## Zip code

Next, let's  explore some of the variables in the business table. We begin by examining the postal code.

### Question 3

#### Question 3a
How are the zip code values stored in python (i.e. data type)?

To answer this you might want to examine a particular entry.

In [12]:
bus.postal_code.sample(5) # NOT SURE THIS IS WHAT THEY ASKED FOR SINCE THEY GAVE TWO LINES TO ANSWER

5177      NaN
2684    94133
715     94114
3801    94109
5839    94121
Name: postal_code, dtype: object

#### Question 3b

What are the unique values of postal_code?

In [13]:
#First way: use unique function on column for postal code

bus['postal_code'].unique()

array(['94104', '94109', '94133', '94110', '94122', '94103', '94115',
       '94131', '94111', '94117', '94107', '94108', '94102', '94132',
       '94105', '94134', '94124', '94116', '94121', '94118', '94112',
       '94123', '94114', '94127', '941102019', '941', '94545', '94066',
       '941033148', nan, '94158', '95105', '94013', '94130', 'CA',
       '92672', '94120', '94143', '94609', '94101', '00000', '94188',
       '94621', '94014', '94129', '94602', 'Ca'], dtype=object)

In [14]:
#Second way: Create a new variable calling unique values of postal code column in bus data frame. (basically same as above..)

unique_postal_code = bus['postal_code'].unique()
print('The unique values in the postal code column are below:')
print(unique_postal_code)

The unique values in the postal code column are below:
['94104' '94109' '94133' '94110' '94122' '94103' '94115' '94131' '94111'
 '94117' '94107' '94108' '94102' '94132' '94105' '94134' '94124' '94116'
 '94121' '94118' '94112' '94123' '94114' '94127' '941102019' '941' '94545'
 '94066' '941033148' nan '94158' '95105' '94013' '94130' 'CA' '92672'
 '94120' '94143' '94609' '94101' '00000' '94188' '94621' '94014' '94129'
 '94602' 'Ca']


#### Question 3c

Let's say we decide to exclude the businesses that have no zipcode for our analysis (which might include food trucks for example).  Use the list of valid 5-digit zip codes below to create a new dataframe called bus_valid, with only businesses whose postal_codes show up in this list of valid zipcodes. How many businesses are there in this new dataframe?

In [15]:
validZip = ["94102", "94103", "94104", "94105", "94107", "94108",
            "94109", "94110", "94111", "94112", "94114", "94115",
            "94116", "94117", "94118", "94121", "94122", "94123", 
            "94124", "94127", "94131", "94132", "94133", "94134"]

In [16]:
#create new df that shows values where postal codes match valid zip values

bus_valid = bus[bus['postal_code'].isin(validZip)]
bus_valid.head()


business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           31      NORMAN'S ICE CREAM AND FREEZES   
4           45                 CHARLIE'S DELI CAFE   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   
4                3202 FOLSOM ST   San Francisco    CA       94110  37.747114   

    longitude  phone_number  
0 -122.403816  +14154217044  
1 -122.421547  +14157763262  
2 -122.403135  +14156779494  
3 -122.419004           NaN  
4 -122.413641  +14156415051

In [17]:
#number of values in the new dataframe

print(len(bus_valid))

5999


## Latitude and Longitude

Another aspect of the data we want to consider is the prevalence of missing values. If many records have missing values then we might be concerned about whether the nonmissing values are representative of the population.

### Question 4
 
Consider the longitude and latitude in the business DataFrame. 

#### Question 4a

How many businesses are missing longitude values, working with only the businesses that are in the list of valid zipcodes?

In [18]:
bus_valid[pd.notnull(bus_valid['longitude'])] #this would print a subset of the df that are valid

#3516 are NOT missing longitude

business_id                                name  \
0              10                    TIRAMISU KITCHEN   
1              19               NRGIZE LIFESTYLE CAFE   
2              24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3              31      NORMAN'S ICE CREAM AND FREEZES   
4              45                 CHARLIE'S DELI CAFE   
...           ...                                 ...   
4025        71784                    URBAN BREAD CAFE   
4029        71823            SUBWAY SANDWICHES #24254   
4031        71841                  RAPHA RACHING, LLC   
4033        71909            SHANGHAI DUMPLING KING 2   
4035        71915              SAPPHIRE ASIAN CUISINE   

                            address           city state postal_code  \
0                     033 BELDEN PL  San Francisco    CA       94104   
1      1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109   
2     500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104   
3              2801 LEAVENWORTH ST   San Francisco    CA       94133   
4                   3202 FOLSOM ST   San Francisco    CA       94110   
...                             ...            ...   ...         ...   
4025                  3901 18TH ST   San Francisco    CA       94114   
4029               200 PINE ST #140  San Francisco    CA       94104   
4031               2198 FILBERT ST   San Francisco    CA       94123   
4033             696 MONTEREY BLVD   San Francisco    CA       94127   
4035             475 SACRAMENTO ST   San Francisco    CA       94111   

       latitude   longitude  phone_number  
0     37.791116 -122.403816  +14154217044  
1     37.786848 -122.421547  +14157763262  
2     37.792888 -122.403135  +14156779494  
3     37.807155 -122.419004           NaN  
4     37.747114 -122.413641  +14156415051  
...         ...         ...           ...  
4025  37.761113 -122.430789  +14155745682  
4029  37.792237 -122.400124  +14152386388  
4031  37.798078 -122.435444  +14152063087  
4033  37.731568 -122.450970  +14152382380  
4035  37.794063 -122.400726  +14153781469  

[3516 rows x 9 columns]

In [19]:
bus_valid[pd.isnull(bus_valid['longitude'])] #this prints a subset onf the ones that are not valid

#2483 ARE missing longitude

business_id                               name             address  \
200           821                       LA SANTANECA    3781 MISSION ST    
210           852  NEW HOLLYWOOD BAKERY & RESTAURANT    652 PACIFIC AVE    
760          2934                   LOU'S FISH SHACK   300 JEFFERSON ST    
826          3123                        LES JOULINS        44 ELLIS ST    
1100         4323                             DIVA'S       1081 POST ST    
...           ...                                ...                 ...   
6310        90234               ACCESS SFUSD THE ARC     1520 HOWARD ST    
6311        90236                     PHO DONG HUONG  667 MONTEREY BLVD    
6312        90237                   WIN'S RESTAURANT    3040 TARAVAL ST    
6313        90238                   USA POWER MARKET     1524 OCEAN AVE    
6314        90239                         ALBA RAY'S    2293 MISSION ST    

               city state postal_code  latitude  longitude  phone_number  
200   San Francisco    CA       94110       NaN        NaN  +14156481034  
210   San Francisco    CA       94133       NaN        NaN  +14153979919  
760   San Francisco    CA       94133       NaN        NaN  +14157715687  
826   San Francisco    CA       94102       NaN        NaN  +14153975397  
1100  San Francisco    CA       94109       NaN        NaN  +14154743482  
...             ...   ...         ...       ...        ...           ...  
6310  San Francisco    CA       94103       NaN        NaN           NaN  
6311  San Francisco    CA       94127       NaN        NaN  +14154945048  
6312  San Francisco    CA       94116       NaN        NaN  +14158606036  
6313  San Francisco    CA       94112       NaN        NaN  +14153331878  
6314  San Francisco    CA       94110       NaN        NaN  +14158881393  

[2483 rows x 9 columns]

In [20]:
no_longitude = bus_valid[pd.isnull(bus_valid['longitude'])] #create a variable that contains only null values

has_longitude = bus_valid[pd.notnull(bus_valid['longitude'])] #create a variable that contains only NON null values

#### Question 4b

Create a new dataframe with one row for each valid zipcode. The dataframe should include the following three columns:

1. `postal_code`: Contains the zip codes in the `validZip` variable above.
2. `null_lon`: The number of businesses in that zipcode with missing `longitude` values.
3. `not_null_lon`: The number of businesses without missing `longitude` values.

In [21]:
bus_valid.head() #checking out the table again...

business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           31      NORMAN'S ICE CREAM AND FREEZES   
4           45                 CHARLIE'S DELI CAFE   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   
4                3202 FOLSOM ST   San Francisco    CA       94110  37.747114   

    longitude  phone_number  
0 -122.403816  +14154217044  
1 -122.421547  +14157763262  
2 -122.403135  +14156779494  
3 -122.419004           NaN  
4 -122.413641  +14156415051

In [22]:
#create first dataframes from the variables we just created and count the name of the variable by postal code 

df_no_longitude = no_longitude[['postal_code','name']].groupby('postal_code').count()


In [23]:
df_no_longitude #test print 

name
postal_code      
94102         196
94103         252
94104          59
94105         100
94107         247
94108          82
94109         146
94110         251
94111         115
94112          59
94114         108
94115          76
94116          35
94117          67
94118         100
94121          55
94122         123
94123          53
94124          88
94127          28
94131          14
94132          65
94133         135
94134          29

In [24]:
#create first dataframes from the variables we just created and count the name of the variable by postal code 

df_has_longitude = has_longitude[['postal_code','name']].groupby('postal_code').count()

In [25]:
df_has_longitude #test print

name
postal_code      
94102         262
94103         306
94104          74
94105         126
94107         204
94108         142
94109         237
94110         334
94111         167
94112         131
94114         122
94115         162
94116          58
94117         130
94118         143
94121         106
94122         147
94123         116
94124          76
94127          43
94131          34
94132          69
94133         282
94134          45

In [26]:
#merge the two dataframes

final_bus = pd.merge(df_no_longitude, df_has_longitude, on='postal_code', how = 'outer')

In [27]:
final_bus #test print

name_x  name_y
postal_code                
94102           196     262
94103           252     306
94104            59      74
94105           100     126
94107           247     204
94108            82     142
94109           146     237
94110           251     334
94111           115     167
94112            59     131
94114           108     122
94115            76     162
94116            35      58
94117            67     130
94118           100     143
94121            55     106
94122           123     147
94123            53     116
94124            88      76
94127            28      43
94131            14      34
94132            65      69
94133           135     282
94134            29      45

In [28]:
#rename the columns

final_bus.columns = ['null_lon', 'not_null_lon']

In [29]:
final_bus

null_lon  not_null_lon
postal_code                        
94102             196           262
94103             252           306
94104              59            74
94105             100           126
94107             247           204
94108              82           142
94109             146           237
94110             251           334
94111             115           167
94112              59           131
94114             108           122
94115              76           162
94116              35            58
94117              67           130
94118             100           143
94121              55           106
94122             123           147
94123              53           116
94124              88            76
94127              28            43
94131              14            34
94132              65            69
94133             135           282
94134              29            45

#### 4c. Do any zip codes appear to have more than their 'fair share' of missing longitude? 

To answer this, you will want to compute the proportion of missing longitude values for each zip code, and print the proportion missing longitude, and print the top five zipcodes in descending order of proportion missing postal_code.


In [30]:
final_bus['sum'] = final_bus[['null_lon', 'not_null_lon']].sum(axis=1) #create a sum column 

final_bus['prop_null_lon'] = final_bus['null_lon']/final_bus['sum'] #create a proportion column 

final_bus #test print

null_lon  not_null_lon  sum  prop_null_lon
postal_code                                            
94102             196           262  458       0.427948
94103             252           306  558       0.451613
94104              59            74  133       0.443609
94105             100           126  226       0.442478
94107             247           204  451       0.547672
94108              82           142  224       0.366071
94109             146           237  383       0.381201
94110             251           334  585       0.429060
94111             115           167  282       0.407801
94112              59           131  190       0.310526
94114             108           122  230       0.469565
94115              76           162  238       0.319328
94116              35            58   93       0.376344
94117              67           130  197       0.340102
94118             100           143  243       0.411523
94121              55           106  161       0.341615
94122             123           147  270       0.455556
94123              53           116  169       0.313609
94124              88            76  164       0.536585
94127              28            43   71       0.394366
94131              14            34   48       0.291667
94132              65            69  134       0.485075
94133             135           282  417       0.323741
94134              29            45   74       0.391892

In [31]:
#sorts in descending order. Probably don't need this given what I ended up doing in my last step...

final_bus.sort_values(by='prop_null_lon', ascending=False)

null_lon  not_null_lon  sum  prop_null_lon
postal_code                                            
94107             247           204  451       0.547672
94124              88            76  164       0.536585
94132              65            69  134       0.485075
94114             108           122  230       0.469565
94122             123           147  270       0.455556
94103             252           306  558       0.451613
94104              59            74  133       0.443609
94105             100           126  226       0.442478
94110             251           334  585       0.429060
94102             196           262  458       0.427948
94118             100           143  243       0.411523
94111             115           167  282       0.407801
94127              28            43   71       0.394366
94134              29            45   74       0.391892
94109             146           237  383       0.381201
94116              35            58   93       0.376344
94108              82           142  224       0.366071
94121              55           106  161       0.341615
94117              67           130  197       0.340102
94133             135           282  417       0.323741
94115              76           162  238       0.319328
94123              53           116  169       0.313609
94112              59           131  190       0.310526
94131              14            34   48       0.291667

In [32]:
#returns five largest

final_bus.nlargest(5,['prop_null_lon'])

null_lon  not_null_lon  sum  prop_null_lon
postal_code                                            
94107             247           204  451       0.547672
94124              88            76  164       0.536585
94132              65            69  134       0.485075
94114             108           122  230       0.469565
94122             123           147  270       0.455556

# Investigate the inspection data

Let's now turn to the inspection DataFrame. Earlier, we found that `ins` has 4 columns, these are named `business_id`, `score`, `date` and `type`.  In this section, we determine the granularity of `ins` and investigate the kinds of information provided for the inspections. 

### Question 5

#### Question 5a
As with the business data, assess whether there is one inspection record for each business, by counting how many rows are in the data and how many unique businesses there are in the data. If they are exactly the same number, it means there is only one inspection per business, clearly.

In [33]:
ins.head() #inspecting dataframe
ins.describe() #counts the number of rows in the dataset

business_id         score          date
count  15430.000000  15430.000000  1.543000e+04
mean   40207.358004     91.179391  2.015071e+07
std    32795.456282      8.024274  8.720966e+03
min       10.000000     46.000000  2.013123e+07
25%     5215.000000     87.000000  2.014072e+07
50%    37773.500000     93.000000  2.015060e+07
75%    71057.000000     98.000000  2.016052e+07
max    89900.000000    100.000000  2.016122e+07

In [34]:
#prints number of unique values for each column
ins.nunique()

business_id    5730
score            49
date            814
type              2
dtype: int64

#### Question 5b

What values does `type` take on? How many occurrences of each value is in the DataFrame? Create a new dataframe named `ins2` by copying `ins` and keeping only records with values of `type` that occur more than 10 times in the original table. In other words, eliminate records that have values of `type` that occur rarely (< 10 times). Check the result to make sure rare types are eliminated.

In [35]:
#shows the two values of type. 1 occurance of complaint and 15429 occurances of routine.

ins['type'].value_counts(ascending=True)

complaint        1
routine      15429
Name: type, dtype: int64

In [36]:
#creates a new variable that ixcludes the type that only occurs once 

routine =['routine']

In [37]:
#Creates new dataframe that includes routine variables from type
ins2 = ins[ins['type'].isin(routine)]

print(ins2)

       business_id  score      date     type
0               10     82  20160503  routine
1               10     94  20140729  routine
2               10     92  20140114  routine
3               19     94  20160513  routine
4               19     94  20141110  routine
...            ...    ...       ...      ...
15425        89515     98  20161103  routine
15426        89701    100  20161206  routine
15427        89790     96  20161129  routine
15428        89803    100  20161215  routine
15429        89900     93  20161206  routine

[15429 rows x 4 columns]


#### Question 5c

Since the data was stored in a .csv file, the dates are formatted as strings such as `20160503`. Once we read in the data, we would like to have dates in an appropriate format for analysis. Add a new column called `year` by capturing the first four characters of the date column. 

Hint: we have seen multiple ways of doing this in class, includings `str` operations, `lambda` functions, `datetime` operations, and others. Choose the method that works best for you :)

In [38]:
#add column 'year' to ins2 dataframe and get the first four string characters of 
ins2['year'] = ins2['date'].astype(str).str[:4]
print(ins2)

       business_id  score      date     type  year
0               10     82  20160503  routine  2016
1               10     94  20140729  routine  2014
2               10     92  20140114  routine  2014
3               19     94  20160513  routine  2016
4               19     94  20141110  routine  2014
...            ...    ...       ...      ...   ...
15425        89515     98  20161103  routine  2016
15426        89701    100  20161206  routine  2016
15427        89790     96  20161129  routine  2016
15428        89803    100  20161215  routine  2016
15429        89900     93  20161206  routine  2016

[15429 rows x 5 columns]


/var/folders/l1/5dp5j9v50g32m73d0hgrnvch0000gn/T/ipykernel_71664/2165242701.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  ins2['year'] = ins2['date'].astype(str).str[:4]


#### Question 5d

What range of years is covered in this data set? Are there roughly same number of inspections each year? Try dropping records for any years with less than 50 inspections and store the result in a new dataframe named `ins3`.

In [39]:
#shows me how many different/unique year values there are
ins2.nunique(axis = 0)

business_id    5730
score            49
date            814
type              1
year              4
dtype: int64

In [40]:
#count values in each column and now we want to get rid of 2013 values b/c it has less than 50.

ins2['year'].value_counts()

2016    5839
2014    5629
2015    3923
2013      38
Name: year, dtype: int64

In [41]:
#create a new variable with valid years

valid_year =["2016", "2015", "2014"]


In [42]:
#create a new df ins3 that only contain years where there are more than 50 values

ins3 = ins2[ins2['year'].isin(valid_year)]
ins3.describe()


business_id         score          date
count  15391.000000  15391.000000  1.539100e+04
mean   40234.948087     91.180755  2.015076e+07
std    32796.883165      8.020365  8.677644e+03
min       10.000000     46.000000  2.014010e+07
25%     5220.000000     87.000000  2.014073e+07
50%    37802.000000     93.000000  2.015061e+07
75%    71076.000000     98.000000  2.016052e+07
max    89900.000000    100.000000  2.016122e+07

In [43]:
#check to make sure i got rid of 2013 values by counting number of values by each year available in df.

ins3['year'].value_counts()

2016    5839
2014    5629
2015    3923
Name: year, dtype: int64

Let's examine only the inspections for one year: 2016. This puts businesses on a more equal footing because [inspection guidelines](https://www.sfdph.org/dph/eh/Food/Inspections.asp) generally refer to how many inspections should occur in a given year.

In [44]:
#creates a new variable with only 2016 values
ins3_2016 = ins3.loc[ins3['year'] == '2016']

print(ins3_2016)


       business_id  score      date     type  year
0               10     82  20160503  routine  2016
3               19     94  20160513  routine  2016
6               24     98  20161005  routine  2016
7               24     96  20160311  routine  2016
10              31     84  20160816  routine  2016
...            ...    ...       ...      ...   ...
15425        89515     98  20161103  routine  2016
15426        89701    100  20161206  routine  2016
15427        89790     96  20161129  routine  2016
15428        89803    100  20161215  routine  2016
15429        89900     93  20161206  routine  2016

[5839 rows x 5 columns]


### Question 6

#### Question 6a

Merge the business and 2016 inspections data, keeping all businesses regardless of whether they show up in the inspections file. Show the first several rows of the resulting dataframe.

In [49]:
#create a merged df by merging two other DF based on ID 

merged_df = pd.merge(bus_valid, ins3_2016, on="business_id", how="outer")
merged_df

business_id                                name  \
0              10                    TIRAMISU KITCHEN   
1              19               NRGIZE LIFESTYLE CAFE   
2              24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3              24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
4              31      NORMAN'S ICE CREAM AND FREEZES   
...           ...                                 ...   
7471        88447                                 NaN   
7472        88524                                 NaN   
7473        88653                                 NaN   
7474        89803                                 NaN   
7475        89900                                 NaN   

                            address           city state postal_code  \
0                     033 BELDEN PL  San Francisco    CA       94104   
1      1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109   
2     500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104   
3     500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104   
4              2801 LEAVENWORTH ST   San Francisco    CA       94133   
...                             ...            ...   ...         ...   
7471                            NaN            NaN   NaN         NaN   
7472                            NaN            NaN   NaN         NaN   
7473                            NaN            NaN   NaN         NaN   
7474                            NaN            NaN   NaN         NaN   
7475                            NaN            NaN   NaN         NaN   

       latitude   longitude  phone_number  score        date     type  year  
0     37.791116 -122.403816  +14154217044   82.0  20160503.0  routine  2016  
1     37.786848 -122.421547  +14157763262   94.0  20160513.0  routine  2016  
2     37.792888 -122.403135  +14156779494   98.0  20161005.0  routine  2016  
3     37.792888 -122.403135  +14156779494   96.0  20160311.0  routine  2016  
4     37.807155 -122.419004           NaN   84.0  20160816.0  routine  2016  
...         ...         ...           ...    ...         ...      ...   ...  
7471        NaN         NaN           NaN   96.0  20160817.0  routine  2016  
7472        NaN         NaN           NaN   96.0  20160927.0  routine  2016  
7473        NaN         NaN           NaN   90.0  20160811.0  routine  2016  
7474        NaN         NaN           NaN  100.0  20161215.0  routine  2016  
7475        NaN         NaN           NaN   93.0  20161206.0  routine  2016  

[7476 rows x 13 columns]

#### Question 6b
Print the 20 lowest rated businesses names, their addresses, and their ratings.

In [50]:
merged_df.nsmallest(20,['score'])

business_id                         name              address  \
4755        71373      GOLDEN RIVER RESTAURANT     5827 GEARY BLVD    
4736        71232                   UNCLE CAFE       65 WAVERLY PL    
6042        81667                 CRAZY PEPPER   2257 SAN JOSE AVE    
7020        87652                    POKI TIME     2101 LOMBARD ST    
2968        34585             HING WANG BAKERY        339 JUDAH ST    
5448        78328                   GOLDEN WOK   295 B ORIZABA AVE    
549          1843          MA'S DIM SUM & CAFE      1315 POWELL ST    
836          2716                 RED A BAKERY      634 CLEMENT ST    
4107        67330      SAN TUNG RESTAURANT LLC      1031 IRVING ST    
4766        71440  NEW GARDEN RESTAURANT, INC.       716 KEARNY ST    
5214        76615                   TOP'S CAFE  2460 SAN BRUNO AVE    
5861        80591    L & G VIETNAMESE SANDWICH         602 EDDY ST    
6386        83647    KOBANI MEDITERANEAN GRILL       3242 SCOTT ST    
624          2044          JOANIE'S DINER INC.   1329 COLUMBUS AVE    
3498        61161                  JADE GARDEN      3608 BALBOA ST    
4130        67564                        ITHAI         720 POST ST    
6657        85587         PAK NAM THAI CUISINE       655 LARKIN ST    
4713        70996             HONG KONG LOUNGE     5322 GEARY BLVD    
4864        72176                 WOLFES LUNCH        1220 16TH ST    
5938        81106          VALLARTA'S TACO BAR     2022 MISSION ST    

               city state postal_code   latitude   longitude  phone_number  \
4755  San Francisco    CA       94121  37.780144 -122.482040  +14156685105   
4736  San Francisco    CA       94108  37.794034 -122.406794  +14155701380   
6042  San Francisco    CA       94112        NaN         NaN  +14157348687   
7020  San Francisco    CA       94123        NaN         NaN  +14156366868   
2968  San Francisco    CA       94122  37.762174 -122.465809  +14156813928   
5448  San Francisco    CA       94132        NaN         NaN  +14155868000   
549   San Francisco    CA       94133  37.796683 -122.410134  +14157883532   
836   San Francisco    CA       94118  37.782888 -122.466101  +14155663888   
4107  San Francisco    CA       94122  37.763891 -122.468805  +14152420828   
4766  San Francisco    CA       94108        NaN         NaN  +14159567888   
5214  San Francisco    CA       94134        NaN         NaN  +14157156868   
5861  San Francisco    CA       94109        NaN         NaN  +14158856402   
6386  San Francisco    CA       94123        NaN         NaN  +14156665567   
624   San Francisco    CA       94133  37.806546 -122.419277  +14158232670   
3498  San Francisco    CA       94121  37.775728 -122.497612  +14152218168   
4130  San Francisco    CA       94109  37.787691 -122.413717  +14156581010   
6657  San Francisco    CA       94109        NaN         NaN  +14156097089   
4713  San Francisco    CA       94121  37.780538 -122.476632           NaN   
4864  San Francisco    CA       94107        NaN         NaN           NaN   
5938  San Francisco    CA       94110        NaN         NaN  +14152528810   

      score        date     type  year  
4755   52.0  20161031.0  routine  2016  
4736   55.0  20160308.0  routine  2016  
6042   55.0  20160901.0  routine  2016  
7020   55.0  20160721.0  routine  2016  
2968   56.0  20160201.0  routine  2016  
5448   56.0  20160824.0  routine  2016  
549    57.0  20160216.0  routine  2016  
836    57.0  20160323.0  routine  2016  
4107   57.0  20160825.0  routine  2016  
4766   57.0  20161121.0  routine  2016  
5214   57.0  20160504.0  routine  2016  
5861   57.0  20160129.0  routine  2016  
6386   58.0  20160714.0  routine  2016  
624    59.0  20160204.0  routine  2016  
3498   59.0  20160107.0  routine  2016  
4130   59.0  20161108.0  routine  2016  
6657   59.0  20160927.0  routine  2016  
4713   60.0  20160121.0  routine  2016  
4864   60.0  20160225.0  routine  2016  
5938   60.0  20160811.0  routine  2016

## Done!

Now commit this notebook to your `hw3_submissions` branch, push it to your GitHub repo, and open a PR!